File: 219-contains-duplicate-ii/contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Map<Integer, List<Integer>> idxs = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (idxs.containsKey(nums[i])) {
                Integer num = nums[i];
                List<Integer> idxList = idxs.get(num);
                for (Integer j : idxList) {
                    boolean val = (Integer) Math.abs(i - j) <= k;
                    if (val) return true; 
                }
                idxList.add(i);
            } else {
                idxs.put(Integer.valueOf(nums[i]), new ArrayList<>());
                idxs.get(Integer.valueOf(nums[i])).add(i);
            }
        }
        return false;
    }
}

File: 219-contains-duplicate-ii/contains-duplicate-ii.js
var containsNearbyDuplicate = function(nums, k) {
    if (k === 0) return false
    let set = new Set();
    let lo = 0, hi = 0;
    while (hi < nums.length) {
        if (set.has(nums[hi])) return true;
        
        if (hi - lo >= k) {
            set.delete(nums[lo]);
            lo++;
        }
        set.add(nums[hi]);
        hi++;
    }
    return false;
};